helix-core/src/test/java/org/apache/helix/controller/rebalancer/waged/constraints/TestPartitionMovementConstraint.java

```diff
@@ -55,126 +57,55 @@ public void init() {
-    score = _constraint.getAssignmentScore(testAssignableNode, _testReplica, _clusterContext);
-    normalizedScore =
-        _constraint.getAssignmentNormalizedScore(testAssignableNode, _testReplica, _clusterContext);
-    // The calculated score is lower than previous value cause the replica's state matches with
-    // best possible is preferred
-    Assert.assertEquals(score, 0.25);
-    Assert.assertEquals(normalizedScore, 0.25);
-
-    // when the replica's allocation matches with baseline only
-    when(testAssignableNode.getInstanceName()).thenReturn(instanceNameC);
-    when(_testReplica.getReplicaState()).thenReturn("Slave");
-    score = _constraint.getAssignmentScore(testAssignableNode, _testReplica, _clusterContext);
-    normalizedScore =
-        _constraint.getAssignmentNormalizedScore(testAssignableNode, _testReplica, _clusterContext);
-    // The calculated score is lower than previous value cause the replica's state matches with
-    // best possible is preferred
-    Assert.assertEquals(score, 0.25);
-    Assert.assertEquals(normalizedScore, 0.25);
-  }
-
-  @Test
-  public void testGetAssignmentScoreWhenBestPossibleMissing() {
-    ResourceAssignment mockResourceAssignment = mock(ResourceAssignment.class);
-    when(mockResourceAssignment.getReplicaMap(new Partition(PARTITION)))
-        .thenReturn(ImmutableMap.of(INSTANCE, "Master"));
-    Map<String, ResourceAssignment> assignmentMap =
-        ImmutableMap.of(RESOURCE, mockResourceAssignment);
-    when(_clusterContext.getBaselineAssignment()).thenReturn(assignmentMap);
-    when(_clusterContext.getBestPossibleAssignment()).thenReturn(Collections.emptyMap());
-    // when the calculated states are both equal to the replica's current state
-    when(_testReplica.getReplicaState()).thenReturn("Master");
-    double score = _constraint.getAssignmentScore(_testNode, _testReplica, _clusterContext);
-    double normalizedScore =
-        _constraint.getAssignmentNormalizedScore(_testNode, _testReplica, _clusterContext);
-
+    score =
+        _bestPossibleStateConstraint.getAssignmentScore(_testNode, _testReplica, _clusterContext);
+    normalizedScore = _bestPossibleStateConstraint
+        .getAssignmentNormalizedScore(_testNode, _testReplica, _clusterContext);
     Assert.assertEquals(score, 1.0);
     Assert.assertEquals(normalizedScore, 1.0);
+
     // when the calculated states are both different from the replica's current state
     when(_testReplica.getReplicaState()).thenReturn("Slave");
-    score = _constraint.getAssignmentScore(_testNode, _testReplica, _clusterContext);
+    score = _baselineConstraint.getAssignmentScore(_testNode, _testReplica, _clusterContext);
     normalizedScore =
-        _constraint.getAssignmentNormalizedScore(_testNode, _testReplica, _clusterContext);
-
+        _baselineConstraint.getAssignmentNormalizedScore(_testNode, _testReplica, _clusterContext);
+    Assert.assertEquals(score, 0.5);
```

Comment:
This PR is fine. But this reminds me of one issue. So if a placement matches both assignment records (but not the state), then this placement would be evaluated as the same score as another candidate with the state, and allocation matches one of the records. In this case, this result will most possibly cause an unnecessary state change. I actually see some flip-flop leader state movements due to this.

So we shall ensure the following 4 levels of scoring at least to avoid this scenario,
1. no match - 0
2. the allocation match one record - less than 0.5, maybe 0.3
3. both state and allocation match - the highest score 1

In this case, if these 2 constraints are configured in the same way, we have 5 levels combined.
0 - no match in both records
0.3 - allocation matches one record
0.6 - allocation matches two records
1.3 - allocation matches one and state matches the other record
2 - state matches both records

There is still a chance that if the user configured the constraint weights so the assumption above is broken again. So I guess we need more work in this area.

Reply:
This needs experiments later, resolve for now.